Simplify function internals by adding new `request_figma()` + `req_figma_query()` helper functions + add URL parsing features

elipousson opened the pull request:
Hello @AleKoure! The package looks super handy and I'm looking forward to working with it. When I took a look at the code, I saw a few opportunities for refactoring to simplify the package internals and reduce duplicative code so I went ahead and made some updates.

I'm hoping this is a welcome contribution but happy to make adjustments if needed. The main additions are:

- `request_figma()` sets up the requests using an "endpoint" string to set the appropriate template for the API call
- `req_figma_query()` is just a wrapper for `httr2::req_url_query()` but also performs the request and returns the JSON body

I also added a new feature allowing users to pass a Figma file URL in place of a file_key (the URL is parsed for a file_key). I also added this functionality for the project_id and I'm using it to make node_id optional. The input checking for the URL is now using cli. I added this to Imports but I figure it was already an upstream dependency so isn't a big lift.

Lastly, I replaced the deprecated call to invoke `rlang::exec()`.

If your up for it, there are a few more changes I'd love to make:

- Replace the checkmate functions with rlang's [standalone type check functions](https://github.com/r-lib/rlang/blob/main/R/standalone-types-check.R) that can be imported with `usethis::use_standalone()` (and any additional custom check functions wrapping `cli::cli_abort()` that may be needed)
- Add a function for setting the Figma API token
- Switch the documentation to Markdown formatting
- Replace `purrr::map_dfr()` and `purrr::imap_dfr()` because all of the row-binding variants are superseded

Let me know what you think!

AleKoure commented:
Hi Elli! Changes look great 🚀 
I agree that the package reached a state that internals need some refactoring!

I'll come back to this ASAP for review, since I was out of office. 

Also, I'll try to open some issues so that things are more transparent since other parts of the package also need improvements at this stage (eg. code repetition, http testing, data tidying, documentation, and grinding deeper in some 'pro' features).

AleKoure commented on R/request_figma.R:
```diff
@@ -0,0 +1,84 @@
+                                   token = NULL,
+                                   ...) {
+  req <- httr2::request(base_url) %>%
+    req_figma_template(
+      endpoint = endpoint,
+      template = template,
+      ...
+    ) %>%
```
This is template is making things complex. Maybe we should not introduce.
The approach looks more generic that it needs to be. 
It shifts the usage of `req_url_path_append()` to pasting strings with `glue()` and introduces an unnecessary ellipses `...`.

elipousson replied:
Hmmm. Under the hood, `httr2::req_template()` is using `paste0()` not glue and the ellipses are how the parameters for each template are passed. I actually really like this pattern since it allows you to see all of the API URL patterns in one place, potentially consolidate parameter checks, and avoid duplication within the public facing functions. I could merge `req_figma_template()` back into `request_figma()` but I think there was at least one function that required the flexibility of two separate functions.

Is your concern with performance? Or maintainability? Happy to explore alternatives if I can get a better understanding of the issue.

AleKoure replied:
Concern is maintenance. Let's continue the discussion here: https://github.com/AleKoure/Rigma/pull/9#discussion_r1234779897

AleKoure commented on R/request_figma.R:
```diff
@@ -0,0 +1,84 @@
+      template = template,
+      ...
+    ) %>%
+    httr2::req_error(
+      body = function(resp) {
+        body <- httr2::resp_body_json(resp)
+        body[["message"]] %||% body[["err"]]
+      }
+    ) %>%
+    req_rigma_agent(token = token)
+
+  if (rlang::is_null(method)) {
+    return(req)
+  }
+
+  req %>% httr2::req_method(method)
+}
```
Great! This is a very nice example of generalization 🚀 
I suggest to keep only this part of the function that is independent of the API at hand. 
We may rename the function to `request_figma_handler()`

```
request_figma_handler <- function(
  req,
  base_url = "https://api.figma.com",
  template = NULL,
  method = NULL,
  token = NULL
) {
  req %>%
    ...
}
```

AleKoure commented on R/thumbnail_color.R:
```diff
@@ -37,6 +39,6 @@ thumbnail_color <- function(path, hex = TRUE) {
   }) %>%
     set_names("red", "green", "blue", "alpha")
-  if (hex) result <- invoke(rgb, result)
+  if (hex) result <- rlang::exec(rgb, result)
```
Great! Just remove the `::` 

```suggestion
  if (hex) result <- exec(rgb, result)
```

AleKoure commented on R/parse_figma_url.R:
```diff
@@ -0,0 +1,129 @@
+#' @noRd
+#' @importFrom rlang check_required is_string
+#' @importFrom cli cli_abort
+set_file_key <- function(file_key, call = caller_env()) {
+  rlang::check_required(file_key, call = call)
+
+  if (is_figma_url(file_key)) {
+    file_key <- parse_url_file_key(file_key)
+  }
+
+  if (is_figma_file_key(file_key)) {
+    return(file_key)
+  }
+
+  cli::cli_abort(
+    "{.arg file_key} must be a file key string or a valid Figma URL
+      starting with {.url https://www.figma.com}",
+    call = call
+  )
+}
```
This PR is quit large and also I wouldn't couple this logic to the API endpoints. Although, it could be handful to expose this as a helper function to parse the URL.

In addition, we should investigate `url_parse()` that potentially could simplify all the logic in this script.

https://httr2.r-lib.org/reference/url_parse.html

I suggest, to stick in this PR to refactoring, and resolve this in a dedicated PR.

AleKoure reviewed (changes requested):
Fantastic approach! Let's stick on this PR to refactoring. It is good practice to keep PRs dedicated to particular tasks for smoother response.

If you proceed with the updates could you please also update the news.md and add yourself as a second author in the description?

elipousson commented on R/get_image.R:
```diff
@@ -55,15 +52,15 @@
 get_image <- function(
   file_key,
-  ids,
+  ids = NULL,
```
The field isn't mandatory if you allow passing a URL as the file_key since a URL can contain a node ID or other ID values.

AleKoure replied:
I agree, this is just a reminder to revert this to mandatory.

elipousson commented on R/delete_comment_reactions.R:
```diff
@@ -33,29 +29,19 @@ delete_comment_reactions <- function(
     emoji
 ) {
-  assert_string(file_key)
+  file_key <- set_file_key(file_key)
```
I think it is reasonable to pull this out for this pull request but I'd really like to see this functionality added to the package overall. I don't see any particular advantage in forcing users to parse file keys and ID values in advance even if a separate utility function is provided.

AleKoure replied:
Yes, the intention is to introduce an R6 class as an abstraction layer on top of this 'low-level' API functions.

If we follow this direction it will be nice to create an R6 object responsible for 'scrapping' FIGMA URLs. Then we can use inheritance and create an interface to the R functions that wrap the endpoints. 

This will also overcome any kitchen sink antipattern, while keeping the inputs of the low level API functions consistent with FIGMA API documentation.

Let's move the discussion to a specific issue!

AleKoure commented on R/request_figma.R:
```diff
@@ -0,0 +1,84 @@
+req_figma_template <- function(req,
+                               endpoint = c("file", "file nodes", "images",
+                                            "image fills", "versions", "users",
+                                            "comments", "comment reactions",
+                                            "projects", "project files",
+                                            "team components", "file components",
+                                            "component", "team component sets",
+                                            "file component set", "component set",
+                                            "team styles", "file styles", "style"),
+                               template = NULL,
+                               ...,
+                               call = rlang::caller_env()) {
+
+  endpoint <- rlang::arg_match(endpoint, error_call = call)
+
+  template <- template %||%
+    switch(endpoint,
+    "file" = "/v1/files/{file_key}",
+    "file nodes" = "/v1/files/{file_key}/nodes",
+    "images" = "/v1/images/{file_key}",
+    "image fills" = "/v1/files/{file_key}/images",
+    "versions" = "/v1/files/{file_key}/versions",
+    "users" = "/v1/me",
+    "comments" = "/v1/files/{file_key}/comments",
+    "comment reactions" = "/v1/files/{file_key}/comments/{comment_id}/reactions",
+    "projects" = "/v1/teams/{team_id}/projects",
+    "project files" = "/v1/projects/{project_id}/files",
+    "team components" = "/v1/teams/{team_id}/components",
+    "file components" = "/v1/files/{file_key}/components",
+    "component" = "/v1/components/{key}",
+    "team component sets" = "/v1/teams/{team_id}/component_sets",
+    "file component set" = "/v1/files/{file_key}/component_sets",
+    "component set" = "/v1/component_sets/{key}",
+    "team styles" = "/v1/teams/{team_id}/styles",
+    "file styles" = "/v1/files/{file_key}/styles",
+    "style" = "/v1/styles/{key}"
+  )
+
+  httr2::req_template(req, template = template, ...)
+}
```
I'm concerned about using ellipsis to pass arguments, although I see that `arg_match()` provides some control. 

Also, pasting with glue is not so defensive and we don't 'fail fast'. `req_url_path_append()` gives a finer grained error response and performs some sanity checks as well.

At least could you please de-couple this from `request_figma_endpoint()`? and if we don't find a way to formalize this better let's keep it.

elipousson replied:
OK – I think I see what you're getting at and I'm happy to separate this from `request_figma_endpoint()` (or at least make it optional). It looks like the currently supported API endpoints only require these four parameters:

- file_key
- comment_id
- key (component key)
- team_id

I could rework this to include named parameters for each and handle validation within the template function so any bad inputs are caught before the API call is attempted. I should be able to spend some time on this over the weekend or next week at the latest. I appreciate you talking this through – it is always helpful to learn how other developers want to see a package structured.

AleKoure reviewed (commented):
Opened separate task  for parsing URLs. We can reconsider adding `parse_figma_url()` there.

elipousson commented:
Thanks for following up on this @AleKoure. I've been a little busy with work but really interested in getting back to this as soon as I can.

elipousson commented:
Moving some scripts around means the requested changes are all flagged as "outdated" but I think all of the requested changes are now incorporated in the pull request. I also removed the row-binding purrr functions to avoid any issues with the planned deprecation. Let me know if you spot anything else that needs to be updated for this pull request and I'll comment separately with my thoughts on the URL parsing when I have a chance!

AleKoure commented on R/get_file_component_sets.R:
```diff
@@ -23,19 +20,14 @@
 #' @export
 get_file_component_sets <- function(file_key) {
-  assert_string(file_key)
+  file_key <- set_file_key(file_key)
```
```suggestion
  assert_file_key(file_key)
```

AleKoure commented on R/get_file_components.R:
```diff
@@ -23,19 +20,14 @@
 #' @export
 get_file_components <- function(file_key) {
-  assert_string(file_key)
+  file_key <- set_file_key(file_key)
```
```suggestion
  assert_file_key(file_key)
```

AleKoure commented on R/get_file_nodes.R:
```diff
@@ -62,13 +59,15 @@
 ) {
-  assert_string(file_key)
+
+  file_key <- set_file_key(file_key)
```
```suggestion
  assert_file_key(file_key)
```

AleKoure commented on R/get_file_styles.R:
```diff
@@ -21,19 +18,14 @@
 #' @export
 get_file_styles <- function(file_key) {
-  assert_string(file_key)
+  file_key <- set_file_key(file_key)
```
```suggestion
  assert_file_key(file_key)
```

AleKoure commented on R/get_file_versions.R:
```diff
@@ -23,19 +20,14 @@
 #' @export
 get_file_versions <- function(file_key) {
-  assert_string(file_key)
+  file_key <- set_file_key(file_key)
```
```suggestion
  assert_file_key(file_key)
```

AleKoure commented on R/get_image.R:
```diff
@@ -55,15 +52,15 @@
 get_image <- function(
   file_key,
-  ids,
+  ids = NULL,
```
Isn't this a mandatory field?

elipousson replied:
Yep. I think I added the NULL default back for one of the parameters that I removed it from but missed this one.

AleKoure commented on R/get_image.R:
```diff
@@ -55,15 +52,15 @@
   version = NULL
 ) {
-  assert_string(file_key)
+  file_key <- set_file_key(file_key)
```
```suggestion
  assert_file_key(file_key)
```

AleKoure commented on R/get_image_fills.R:
```diff
@@ -28,19 +25,14 @@
 #' @export
 get_image_fills <- function(file_key) {
-  assert_string(file_key)
+  file_key <- set_file_key(file_key)
```
```suggestion
  assert_file_key(file_key)
```

AleKoure commented on R/post_comment.R:
```diff
@@ -38,29 +37,29 @@ post_comment <- function(
   client_meta
 ) {
-  assert_string(file_key)
+  file_key <- set_file_key(file_key)
```
```suggestion
  assert_file_key(file_key)
```

AleKoure commented on R/post_comment_reactions.R:
```diff
@@ -33,25 +32,20 @@ post_comment_reactions <- function(
     emoji
 ) {
-  assert_string(file_key)
+  file_key <- set_file_key(file_key)
```
```suggestion
  assert_file_key(file_key)
```

AleKoure reviewed (commented):
Great job!

Added some minor comments, and we are ready to merge.

We should start some more testing, and maybe improve a little bit on unit testing, before proceeding to the next CRAN release, since we have some major changes in this PR.

Also, please update the news.md, ensure that checkmate the namespace is OK for the modified scripts (e.g. I think checkmate is no longer needed in some scripts since we replaced with `assert_file_ke()`) with some bullets for the things you addressed in this PR, and add yourself as a second author in the description!

elipousson commented on R/text_data_from_styles.R:
```diff
@@ -34,14 +34,15 @@ text_data_from_styles <- function(design_tibble) {
     get_file_nodes(file_key, .) %>%
     chuck("nodes") %>%
-    imap_dfr(~ append(
+    map(~ append(
```
Yep. This was just a mistake on my part – glad you caught it.

AleKoure reviewed (approved):
Approved! The token is missing but that's OK. 
There will be some additional tweaks before the next CRAN release

elipousson commented:
Sounds great. I restored the `*_dfr` functions after struggling to get everything working and realizing there was no good reason to include them in this push request if it wasn't easy. I'll give everything a double-check over the weekend, add (or suggest the addition of) tests if needed, and (hopefully) merge.
